src/effects/backends/builtin/metronomeeffect.cpp

```diff
@@ -57,71 +58,87 @@ void MetronomeEffect::processChannel(
-            }
+    if (enableState == EffectEnableState::Enabling) {
+        if (m_pSyncParameter->toBool() && groupFeatures.beat_fraction_buffer_end.has_value()) {
+            // Skip first click and sync phase
+            gs->m_framesSinceClickStart = clickSize;
+        } else {
+            // click right away after enabling
+            gs->m_framesSinceClickStart = 0;
         }
-    } else {
-        maxFrames = static_cast<SINT>(
-                engineParameters.sampleRate() * 60 / m_pBpmParameter->value());
     }
 
-    SampleUtil::copy(pOutput, pInput, engineParameters.samplesPerBuffer());
-
     if (gs->m_framesSinceClickStart < clickSize) {
-        // still in click region, write remaining click frames.
+        // In click region, write remaining click frames.
         const SINT copyFrames =
                 math_min(engineParameters.framesPerBuffer(),
                         clickSize - gs->m_framesSinceClickStart);
         SampleUtil::addMonoToStereo(pOutput, &click[gs->m_framesSinceClickStart], copyFrames);
     }
 
-    gs->m_framesSinceClickStart += engineParameters.framesPerBuffer();
+    double bufferEnd = gs->m_framesSinceClickStart + engineParameters.framesPerBuffer();
+
+    double nextClickStart = bufferEnd; // default to "no new click";
+    if (m_pSyncParameter->toBool() && groupFeatures.beat_length_frames.has_value()) {
+        if (groupFeatures.beat_fraction_buffer_end.has_value()) {
+            double beatLength = *groupFeatures.beat_length_frames;
+            if (groupFeatures.scratch_rate.has_value()) {
+                if (*groupFeatures.scratch_rate != 0.0) {
+                    beatLength /= *groupFeatures.scratch_rate;
+                } else {
+                    // no transport, nothing to do.
+                    return;
+                }
+            }
 
-    if (gs->m_framesSinceClickStart > maxFrames) {
-        // overflow, all overflowed frames are the start of a new click sound
-        gs->m_framesSinceClickStart -= maxFrames;
-        while (gs->m_framesSinceClickStart > engineParameters.framesPerBuffer()) {
-            // loop into a valid region, this happens if the maxFrames was lowered
-            gs->m_framesSinceClickStart -= engineParameters.framesPerBuffer();
+            double beatToBufferEnd;
+            if (beatLength > 0) {
+                beatToBufferEnd =
+                        beatLength *
+                        *groupFeatures.beat_fraction_buffer_end;
+            } else {
+                beatToBufferEnd =
+                        beatLength * -1 *
+                        (1 - *groupFeatures.beat_fraction_buffer_end);
+            }
+
+            if (bufferEnd > beatToBufferEnd) {
+                // We have a new beat before the current buffer ends
+                nextClickStart = bufferEnd - beatToBufferEnd;
+            }
+        } else {
+            nextClickStart = *groupFeatures.beat_length_frames;
```

Comment:
Can this actually happen?

Reply:
No, I have added it only for completeness. We may consider to use both in a optional struct to get around it.

Reply:
do you mean `beat_length_frames` and `beat_fraction_buffer_end` or `beat_fraction_buffer_end` and `scratch_rate`? because according to `BpmControl::collectFeatures` if `rate_ratio` is 0, we can get a `beat_fraction_buffer_end` but not a `beat_length_frames`

Reply:
We always have it all in reality, because we do not allow a rate of 0 but it is possible via writing the CO. 
We should probably treat this case in the same way we treat scratch = 0

Reply:
yes, I agree. I implemented that in a follow up. will push shortly.

Reply:
I am already on it.

Reply:
sure that works, but it doesn't eliminate the original case I asked about (which is when `beat_length_frames.has_value()` but `!beat_fraction_buffer_end.has_value()`).

Reply:
This can not happen. But looking at the API it, can. `beat_fraction_buffer_end.has_value() == false` means the track has no value. In this case `beat_length_frames.has_value()` is always false.

Reply:
right, so that should be reflected in `GroupFeatureState`, do you agree?

Reply:
I am not sure, I because we may consider a sound source that has bpm info but no phase info.